File: Libraries/Graphics2DText/resources/csharp/Grahpics2DTextHelper.cs
using System;

internal static class Graphics2DTextHelper
{
    public static object CreateNativeFont(int fontType, int fontClass, string fontId, int size, bool isBold, bool isItalic)
    {
        if (fontType == 1) // embedded font resource
        {
            return Interpreter.UniversalFont.FromResources(fontId, size, isBold, isItalic);
        }
        else if (fontType == 3) // system font
        {
            return Interpreter.UniversalFont.FromSystem(fontId, size, isBold, isItalic);
        }
        else
        {
            throw new NotImplementedException("Not implemented.");
        }
    }

    public static bool IsSystemFontAvailable(String name)
    {
        // TODO: call this function directly
        return Interpreter.UniversalFont.IsSystemFontAvailable(name);
    }

    public static Interpreter.UniversalBitmap RenderCharTile(object nativeFont, int charId, int[] sizeOut)
    {
        Interpreter.UniversalBitmap bmp = new Interpreter.UniversalBitmap((Interpreter.UniversalFont)nativeFont, (char)charId);
        sizeOut[0] = bmp.Width;
        sizeOut[1] = bmp.Height;

        double leftMarginRatio = bmp.IsCairo ? 0.0 : 0.125;
        double widthRatio = bmp.IsCairo ? 1.0 : 0.55;
        sizeOut[2] = (int)(bmp.Width * leftMarginRatio);
        sizeOut[3] = (int)(bmp.Width * widthRatio);

        return bmp;
    }

    public static Interpreter.UniversalBitmap GenerateTextureAndAllocateFloatInfo(object[][] nativeDatas, int[] coordinateInfo, int width, int height)
    {
        Interpreter.UniversalBitmap output = new Interpreter.UniversalBitmap(width, height);
        Interpreter.UniversalBitmap.DrawingSession g = output.GetActiveDrawingSession();
        int length = nativeDatas.Length;
        int tileX, tileY, tileWidth, tileHeight;
        object[] nativeData;
        for (int i = 0; i < length; ++i)
        {
            nativeData = nativeDatas[i];
            tileX = coordinateInfo[i * 4];
            tileY = coordinateInfo[i * 4 + 1];
            tileWidth = coordinateInfo[i * 4 + 2];
            tileHeight = coordinateInfo[i * 4 + 3];

            g.Draw((Interpreter.UniversalBitmap)nativeData[0], tileX, tileY, 0, 0, tileWidth, tileHeight);
			nativeData[5] = -1;
			nativeData[6] = null;
            nativeData[7] = tileX;
            nativeData[8] = tileY;
            nativeData[9] = tileX + tileWidth;
            nativeData[10] = tileY + tileHeight;

            nativeData[11] = output.Width;
            nativeData[12] = output.Height;
        }
        g.Flush();
        return output;
    }

    public static int LoadOpenGlTexture(object bitmapObj)
    {
        Interpreter.UniversalBitmap bitmap = (Interpreter.UniversalBitmap)bitmapObj;
        return Interpreter.Libraries.Game.GlUtil.ForceLoadTexture(bitmap);
    }
}

File: Libraries/Graphics2DText/resources/java/Grahpics2DTextHelper.java
package org.crayonlang.libraries.graphics2dtext;

import java.awt.Color;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.font.TextAttribute;
import java.awt.Graphics2D;
import java.awt.GraphicsEnvironment;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.HashMap;
import org.crayonlang.interpreter.structs.*;
import org.crayonlang.interpreter.ResourceReader;

final class Graphics2DTextHelper {
	
	private Graphics2DTextHelper() { }
	
	public static Object createNativeFont(int fontType, int fontClass, String fontId, int size, boolean isBold, boolean isItalic) {
		int style = Font.PLAIN;
		if (isBold) style |= Font.BOLD;
		if (isItalic) style |= Font.ITALIC;
		
		int adjustedSize = size * 3 / 2;
		Font font;
		if (fontType == 1) { // font embedded resource
			font = ResourceReader.getFontResource(fontId, adjustedSize, style);
		} else if (fontType == 3) { // system font
			font = new Font(fontId, style, adjustedSize);
		} else {
			throw new RuntimeException("Not implemented.");
		}
		
		// adjust kerning
		HashMap<TextAttribute, Object> attributes = new HashMap<TextAttribute, Object>();
		attributes.put(TextAttribute.TRACKING, -0.05);
		font = font.deriveFont(attributes);
		
		return font;
	}
	
	public static void addAll(ArrayList<Value> values, Value... toAdd) {
		for (Value value : toAdd) {
			values.add(value);
		}
	}
	
	public static boolean isSystemFontAvailable(String name) {
		String[] fontNames = GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
		for (String fontName : fontNames) {
			if (fontName.equals(name)) {
				return true;
			}
		}
		return false;
	}
	
	private static final BufferedImage DUMMY_IMAGE = new BufferedImage(1, 1, BufferedImage.TYPE_INT_ARGB);
	
	public static Object renderText(
		int[] sizeOut,
		Object fontObj,
		int red,
		int green,
		int blue,
		String text) {
		
		Font font = (Font) fontObj;
		FontMetrics fontMetrics = DUMMY_IMAGE.createGraphics().getFontMetrics(font);
		int baselineHeight = fontMetrics.getHeight();
		int height = baselineHeight * 3 / 2;
		int margin = height / 32;
		int width = fontMetrics.stringWidth(text) + margin * 2 + 4;
		BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
		Graphics2D g = image.createGraphics();
		Color color = new Color(red, green, blue);
		g.setColor(color);
		g.setFont(font);
		g.drawString(text, margin, baselineHeight);
		sizeOut[0] = width;
		sizeOut[1] = height;
		return image;
	}
}
